src/core/segment_reader.rs

```diff
@@ -330,6 +447,75 @@ impl SegmentReader {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// FieldMetadata
+pub struct FieldMetadata {
+    /// The field name
+    pub field_name: String,
+    /// The field type
+    pub typ: Type,
+    /// Is the field indexed for search
+    pub indexed: bool,
+    /// Is the field stored in the doc store
+    pub stored: bool,
+    /// Is the field stored in the columnar storage
+    pub fast: bool,
+}
+impl PartialOrd for FieldMetadata {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for FieldMetadata {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.field_name.cmp(&other.field_name) {
+            Ordering::Equal => self.typ.cmp(&other.typ),
+            other => other,
+        }
+    }
+}
+
+impl BitOrAssign for FieldMetadata {
+    fn bitor_assign(&mut self, rhs: Self) {
+        assert!(self.field_name == rhs.field_name);
+        assert!(self.typ == rhs.typ);
+        self.indexed |= rhs.indexed;
+        self.stored |= rhs.stored;
+        self.fast |= rhs.fast;
+    }
+}
+
+/// Helper to merge the field metadata from multiple segments.
+pub fn merge_field_meta_data(
+    input: Vec<Vec<FieldMetadata>>,
+    schema: &Schema,
+) -> Vec<FieldMetadata> {
+    // Maybe too slow for the high cardinality case
+    fn is_field_stored(field_name: &str, schema: &Schema) -> bool {
+        schema
+            .find_field(field_name)
+            .map(|(field, _path)| schema.get_field_entry(field).is_stored())
+            .unwrap_or(false)
+    }
+
+    let mut out = Vec::new();
+    for (_key, mut group) in &input
+        .into_iter()
+        .kmerge_by(|a, b| a < b)
```

Comment:
left / right
```suggestion
        .kmerge_by(|a, b| a < b)
```

Reply:
Can you. remove the Ord implementation of FieldMetadata, create a function extracting the key you want (returning the pair field name field type ).
Put that in that file, and call it here instead?

Reply:
I prefer to have the Ord implementation, so it will behave consistent with `sort_by` etc.